Support launch_ipdb_on_exception as a function decorator

alanbernstein opened the pull request:
Fixes #215.

I chose `iex` (**i**pdb **ex**ception) instead of my previous conflicting `pm`. It's a concise name because that's how I like my debugging tools (and I assume the existing `pm` was also chosen for conciseness). If that's not acceptable here, I think I'll instead try to figure out a good way to make `launch_ipdb_on_exception` itself a function decorator, so it can be imported as `from ipdb import launch_ipdb_on_exception as iex`.

gotcha commented:
Thanks for the PR. Is there a reason it is in draft state ?

Would you mind adding a commit which fixes `decorator` dependency for Python 3.4 in `setup.py` ?

Turn

      ':python_version == "3.4"': ['ipython >= 6.0.0, < 7.0.0', 'toml >= 0.10.2'],

into

      ':python_version == "3.4"': ['ipython >= 6.0.0, < 7.0.0', 'toml >= 0.10.2', 'decorator < 5'],

I think it should fix CI.

alanbernstein commented:
> Thanks for the PR. Is there a reason it is in draft state ?

I'm happy to contribute, thanks for being welcoming. I didn't understand the version-specific dependency issue with 3.4, I had planned to look into it over the weekend.

> Would you mind adding a commit which fixes `decorator` dependency for Python 3.4 in `setup.py` ?

Added.

gotcha commented:
> I suppose the addition of Decontext class is pulling down the coverage - maybe it makes more sense to just use the decorator package?

As `decorator` is a dependency of `IPython` since 4.0.0 and we depend on later versions of `IPython` for all Python versions, I think you are right that it would make sense to use `decorator`.

Can you fix that point ?

When specifying the dependency in `setup.py`, can you add a comment explaining that `decorator` is also a dependency of `Ipython` ?

alanbernstein commented:
Sorry to leave this hanging, this was a little more time consuming as I had to refresh my memory of how decorators and context managers interact here. I wasn't aware that the `decorator` library provides a replacement `contextmanager` that does just the right thing here (https://github.com/micheles/decorator/blob/master/docs/documentation.md#contextmanager).

I added a simple comment to `setup.py`, but I don't fully grok the dependency on decorator, I may have misunderstood what you asked for.

I also corrected some unrelated formatting issues in README.rst, would you prefer a different PR for that?

gotcha commented:
Released in https://pypi.org/project/ipdb/0.13.8/
